Added gradient hover view with save/cancel buttons in EditProfileVC

caitlynjin opened the pull request:
## Overview

Added a gradient hover view along with buttons to cancel or save the changed data fields in `EditProfileViewController` to the database and `ProfileViewController`. Created delete account button inside the scroll view (not yet functional).



## Changes Made

### Edit Profile Page

- Created a cancel button that simply pops the view controller and a save button that updates the user data and in `ProfileViewController` based on input information from the page.
  - Save button has the same implementation as `viewWillDisappear`, which was previously used to save the user data.
  - Removed `viewWillDisappear`.
- Created a white gradient that fades to transparent, which sticks to the screen (not scrollable) underneath the buttons.
  - Made user unable to interact with `gradientView`, so the user will be able to scroll underneath the gradient.
- Changed constraints regarding hiding/unhiding `phoneNumTextField` because the gradient would overlap with the view. The screen updates according to the button selection.
- Created the UI for the delete account button.
- Moved networking call to separate section under MARK: - Requests.
- Factored out constants in file into a struct `Constants`.
- Created functions to dismiss the keyboard display from the text field after interacting with the slider. Keyboard still displays when tapping or scrolling the view.



### Minor Changes

- Made back button and navigation bar hidden in pushed instance of `EditProfileViewController`.
- Added fonts in `Extension+UIFont.swift`. Changed SF Pro fonts to `.systemFont`.
- Reorganized and added colors in `Extension+UIColor.swift`.



## Test Coverage

- Did manual testing through simulator to ensure view constraints after button selection is implemented correctly.
- Did manual testing for dismissing the keyboard, ensuring that it is still scrollable while visible, but hidden while using the slider views.


## Next Steps

- Create networking request for deleting a user account and call the request when selecting the delete button (add confirmation message?).
- Double check save/cancel button after our backend is working.



## Screenshots

<details>

  <summary>Before</summary>

  https://github.com/cuappdev/scoop-ios/assets/118781810/4a3e0f8f-450c-4cf8-95a5-1dfb2d96b4d9

</details>

<details>

  <summary>After</summary>

  https://github.com/cuappdev/scoop-ios/assets/118781810/a017e7e8-d9a5-4e57-bf3e-4389279d9499

</details>

<details>

  <summary>After v2</summary>

  https://github.com/cuappdev/scoop-ios/assets/118781810/397ba542-1f68-4869-8205-e2e34cb5b6c7

</details>

vinnie4k commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -340,8 +345,10 @@ class EditProfileViewController: UIViewController {
 
             if emailButton.isSelected {
                 phoneNumTextField.isHidden = true
+                view.layoutIfNeeded()
```
You can move this outside of the if statement since it is being used in both cases.

vinnie4k commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,81 @@ class EditProfileViewController: UIViewController {
+    }
+    
+    private func setupGradientView() {
+        gradientLayer.colors = [UIColor.white.withAlphaComponent(0).cgColor, UIColor.white.withAlphaComponent(0.3).cgColor, UIColor.white.withAlphaComponent(0.6).cgColor, UIColor.white.withAlphaComponent(0.9).cgColor, UIColor.white.cgColor, UIColor.white.cgColor, UIColor.white.cgColor, UIColor.white.cgColor, UIColor.white.cgColor, UIColor.white.cgColor, UIColor.white.cgColor, UIColor.white.cgColo ...
```
This is difficult to read. Whenever you have an array with many values, separate each element with a new line.

vinnie4k commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,81 @@ class EditProfileViewController: UIViewController {
+    @objc func saveEdit() {
+        updateUser()
+        self.navigationController?.popViewController(animated: true)
+    }
```
If these are not being called anywhere else, mark them as `private`.

vinnie4k commented on Scoop/Supporting/Utilities/Extension+UIColor.swift:
```diff
@@ -9,18 +9,21 @@ import Foundation
+    static let scoopDarkGreen = UIColor(red: 58/255, green: 146/255, blue: 117/255, alpha: 1)
+    static let secondaryGreen = UIColor(red: 206/255, green: 233/255, blue: 220/255, alpha: 1)
+    static let skipButtonColor = UIColor(red: 151/255, green: 151/255, blue: 151/255, alpha: 1)
+    static let textFieldBorderColor = UIColor(red: 106/255, green: 115/255, blue: 125/255, alpha: 1)
```
Nice.

vinnie4k commented on Scoop/Supporting/Utilities/Extension+UIFont.swift:
```diff
@@ -11,7 +11,9 @@ import UIKit
     
     static let flowHeader = UIFont(name: "Sen-Regular", size: 24)
-    static let bodyNormal = UIFont(name: "SFPro", size: 16)
+    static let bodyBold = UIFont.systemFont(ofSize: 16, weight: .bold)
+    static let bodyNormal = UIFont.systemFont(ofSize: 16)
+    static let bodySemibold = UIFont.systemFont(ofSize: 16, weight: .semibold)
```
Really nice. Using extensions is very powerful.

vinnie4k reviewed (changes requested):
Great work! Just one think you might want to double check, usually you can dismiss a ViewController by swiping from the side. Since there is no longer a back button, I would double check that this feature is disabled. The only way to dismiss the view would either be to save or to cancel.

Please address my comments then request me again.

caitlynjin commented:
> Great work! I really like the PR description. Just one think you might want to double check, usually you can dismiss a ViewController by swiping from the side. Since there is no longer a back button, I would double check that this feature is disabled. The only way to dismiss the view would either be to save or to cancel.
> 
> Please address my comments then request me again.

It looks like it's already disabled 👍

rs929 commented on Scoop/Controller/LoginViewController.swift:
```diff
@@ -58,7 +58,10 @@ class LoginViewController: UIViewController {
         
         let signInAction = UIAction { action in
-            self.signIn()
+//            self.signIn()
```
Uncomment this, since this is only temporary while the backend is down

rs929 commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+
+        deleteButton.snp.makeConstraints { make in
+            make.leading.trailing.equalToSuperview().inset(40)
+            make.height.equalTo(51)
```
I would just make this 50 to keep it even

tiffany-pann replied:
Hmm does this look okay for smaller screens? i.e. iPhone SE lol. If np, feel free to disregard my following comments

rs929 commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+        
+        gradientView.snp.makeConstraints { make in
+            make.bottom.width.equalToSuperview()
+            make.height.equalTo(201)
```
Same as above, make this 200

rs929 commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+    }
+    
+    private func setupGradientView() {
+        gradientLayer.colors = [
```
Maybe add this to the UIColor Extension, in case we need to utilize the same type of gradient again

rs929 commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+        cancelButton.backgroundColor = UIColor.white
+        cancelButton.layer.borderColor = UIColor.systemGray2.cgColor
+        cancelButton.layer.borderWidth = 1
+        cancelButton.layer.cornerRadius = 24
```
Make this 25 to keep it consistent with other buttons

rs929 commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+            make.leading.equalToSuperview().inset(24)
+            make.bottom.equalTo(view.safeAreaLayoutGuide.snp.bottom).inset(39)
+            make.width.equalTo(102)
+            make.height.equalTo(51)
```
Make this 50

tiffany-pann replied:
Similarly, same thing as previous comment

rs929 commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+        saveButton.setTitleColor(UIColor.offBlack, for: .normal)
+        saveButton.titleLabel?.font = UIFont.bodyBold
+        saveButton.backgroundColor = UIColor.secondaryGreen
+        saveButton.layer.cornerRadius = 24
```
Same as above

tiffany-pann commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+            UIColor.white.withAlphaComponent(0.3).cgColor,
+            UIColor.white.withAlphaComponent(0.6).cgColor,
+            UIColor.white.withAlphaComponent(0.9).cgColor,
+            UIColor.white.cgColor,
```
so much white ⬜️🤍

caitlynjin replied:
indeed 🤍

rs929 commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+            make.trailing.equalToSuperview().inset(24)
+            make.bottom.equalTo(view.safeAreaLayoutGuide.snp.bottom).inset(39)
+            make.width.equalTo(102)
+            make.height.equalTo(51)
```
Same as above

rs929 commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+        
+        cancelButton.snp.makeConstraints { make in
+            make.leading.equalToSuperview().inset(24)
+            make.bottom.equalTo(view.safeAreaLayoutGuide.snp.bottom).inset(39)
```
Make this 40

rs929 commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+        
+        saveButton.snp.makeConstraints { make in
+            make.trailing.equalToSuperview().inset(24)
+            make.bottom.equalTo(view.safeAreaLayoutGuide.snp.bottom).inset(39)
```
Make this 40

tiffany-pann commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+        cancelButton.snp.makeConstraints { make in
+            make.leading.equalToSuperview().inset(24)
+            make.bottom.equalTo(view.safeAreaLayoutGuide.snp.bottom).inset(39)
+            make.width.equalTo(102)
```
Same thing as prev. concern about this being a constant

tiffany-pann commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -561,8 +567,95 @@ class EditProfileViewController: UIViewController {
+        saveButton.snp.makeConstraints { make in
+            make.trailing.equalToSuperview().inset(24)
+            make.bottom.equalTo(view.safeAreaLayoutGuide.snp.bottom).inset(39)
+            make.width.equalTo(102)
```
Same concern about this being a constant & its height constraint as well

rs929 reviewed (approved):
Looks good to me! Just a few minor changes. Whenever there are any constraints that are odd or not a multiple of 4, just round or truncate to keep things consistent.

tiffany-pann reviewed (changes requested):
Nice PR! Feel free to re-request after the small constants issue. From your screen-recording, I noticed that the buttons felt a little too high up, not sure if this was based on design (but if it is, ignore me). 

Also, this is something else really small, but from the "Before" screen recording, I noticed that the phone number text-field still has its cursor (little vertical bar) flashing even after you're finished inputting and moved on to interacting with the sliders. This shouldn't be too bad, I think you could just add something like: 
`func textFieldDidEndEditing(_ textField: UITextField) {
        textField.tintColor = .clear
    }
`
No worries about this tho, just something small I noticed! Great work so far ( •̀ ω •́ )✧

vinnie4k commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -59,7 +59,12 @@ class EditProfileViewController: UIViewController {
     // MARK: - Constants
-    
+
+    private let bottomInset = 36
+    private let buttonCornerRadius = CGFloat(25)
+    private let buttonHeight = 50
+    private let buttonLeadingTrailingInset = 24
+    private let buttonWidth = 104
     private let leadingTrailingInset = 32
     private let textFieldHeight = 56
```
If this style is used in other places throughout the codebase then it's fine. However, I like to convert it to the following:
```suggestion
    private struct Constants {
        static let bottomInset: CGFloat = 36
        static let buttonCornerRadius: CGFloat = CGFloat(25)
        static let buttonHeight: CGFloat = 50
        static let buttonLeadingTrailingInset: CGFloat = 24
        static let buttonWidth: CGFloat = 104
        static let leadingTrailingInset: CGFloat = 32
        static let textFieldHeight: CGFloat = 56
    }
```
You would then call it using `Constants.bottomInset`

vinnie4k commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -642,6 +656,22 @@ class EditProfileViewController: UIViewController {
+        view.endEditing(true)
+    }
+
+    func dismissKeyboardFromTap() {
```
Is this being called anywhere else? Mark this as `private` if so.

vinnie4k commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -642,6 +656,22 @@ class EditProfileViewController: UIViewController {
+        view.addGestureRecognizer(tap)
+    }
+
+    func dismissKeyboardFromPan() {
```
Is this being called anywhere else? Mark this as `private` if so.

vinnie4k commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -642,6 +656,22 @@ class EditProfileViewController: UIViewController {
+    }
+
+    func dismissKeyboardFromPan() {
+        let pan: UIPanGestureRecognizer = UIPanGestureRecognizer(target: self, action: #selector(dismissKeyboardTouchOutside))
```
```suggestion
        let pan = UIPanGestureRecognizer(target: self, action: #selector(dismissKeyboardTouchOutside))
```

vinnie4k commented on Scoop/Controller/Profile/EditProfileViewController.swift:
```diff
@@ -642,6 +656,22 @@ class EditProfileViewController: UIViewController {
+    }
+
+    func dismissKeyboardFromTap() {
+        let tap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(dismissKeyboardTouchOutside))
```
```suggestion
        let tap = UITapGestureRecognizer(target: self, action: #selector(dismissKeyboardTouchOutside))
```

vinnie4k reviewed (approved):
Address my comments and you should be good to go!

caitlynjin commented:
> Nice PR! Feel free to re-request after the small constants issue. From your screen-recording, I noticed that the buttons felt a little too high up, not sure if this was based on design (but if it is, ignore me).
> 
> Also, this is something else really small, but from the "Before" screen recording, I noticed that the phone number text-field still has its cursor (little vertical bar) flashing even after you're finished inputting and moved on to interacting with the sliders. This shouldn't be too bad, I think you could just add something like: `func textFieldDidEndEditing(_ textField: UITextField) { textField.tintColor = .clear } ` No worries about this tho, just something small I noticed! Great work so far ( •̀ ω •́ )✧

I tried the textFieldDidEndEditing, but I changed it to dismissing the keyboard for just the sliders because the keyboard doesn't need to be displayed for them. I made an updated video in the PR!
